### PR TITLE
Makara

### PR DESCRIPTION
**Pull request by pleary**
- Adds makara gem which can be activated by updating database.yml
- Currently opt-in:
  - Delayed jobs are allow-listed in a plugin
  - Controller actions need to `prepend_around_action :enable_replica`
- Some overrides like assuming all DJ queries to need the primary, and allowing for context resetting in a non-Rack environment

### CODE REVIEW

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -291,6 +291,15 @@ def logstash_catchall
...
   end
 
+  def enable_replica
+    begin
+      ActiveRecord::Base.connection.enable_replica
+      yield
+    ensure
+      ActiveRecord::Base.connection.disable_replica
+    end
+  end
```
Comment (kueda) [inline app/controllers/application_controller.rb:301]: Just so I understand, is this turning on read/write splitting? I.e. if an action has this around it, Makara will perform SELECTs in the replica and INSERTs in the primary?

File: app/controllers/application_controller.rb [hunk 2]
```diff
@@ -745,7 +754,9 @@ def user_viewed_updates_for( record, options = {} )
     return unless logged_in?
     updates = UpdateAction.where( resource: record )
     if options[:delay]
-      UpdateAction.delay( priority: USER_PRIORITY ).user_viewed_updates( updates, current_user.id )
+      ActiveRecord::Base.connection.without_sticking do
```
Comment (kueda) [inline app/controllers/application_controller.rb:757]: Again, just for my edification, this means that inside this block, if Makara directs a query to the primary, that does not mean that all subsequent queries get sent to the primary, which is the default "sticky" behavior, right?

File: config/initializers/makara.rb [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+      SQL_SKIP_STICKINESS_MATCHERS = [] unless defined?( SQL_SKIP_STICKINESS_MATCHERS )
+      CUSTOM_SQL_SKIP_STICKINESS_MATCHERS  = SQL_SKIP_STICKINESS_MATCHERS + [/FROM "sessions"/i].map(&:freeze).freeze
+
+      def sql_primary_matchers
+        CUSTOM_SQL_PRIMARY_MATCHERS
+      end
+
+      def sql_skip_stickiness_matchers
+        CUSTOM_SQL_SKIP_STICKINESS_MATCHERS
+      end
```
Comment (kueda) [inline config/initializers/makara.rb:49]: This makes sure the module is defined, but what are these methods doing?

File: lib/elastic_model/acts_as_elastic_model.rb [hunk 4]
```diff
@@ -259,7 +259,9 @@ def bulk_index(batch, options = { })
             refresh: options[:wait_for_index_refresh] ? "wait_for" : false
           })
           if batch && batch.length > 0 && batch.first.respond_to?(:last_indexed_at)
-            where(id: batch).update_all(last_indexed_at: Time.now)
+            ActiveRecord::Base.connection.without_sticking do
+              where(id: batch).update_all(last_indexed_at: Time.now)
+            end
```
Comment (kueda) [inline lib/elastic_model/acts_as_elastic_model.rb:264]: Presumably the UPDATE gets run on the primary, but what are the subsequent queries that need to be "unstuck" and run on the replica?

**Review by kueda — COMMENTED** [review]
I don't think you need my approval for anything here, I'm mostly reviewing so I can learn about how this works. On additional question: I take it chef will handle the changes Makara needs in database.yml?

File: app/controllers/application_controller.rb (hunk 1 above)
Comment (pleary) [reply app/controllers/application_controller.rb:301]: Yes. This is a toggle I added for temporary use where we can opt-in a request to read/write splitting. If a controller/action does not call this with `prepend_around_action` then all its queries will run against the primary. If it does call this, it'll use read/write splitting that I plan to configure with "stickiness". Meaning all reads to to the replica until there's an INSERT or UPDATE or anything else that is known to need the primary. After a query goes to the primary (unless we temporarily ignore stickiness), all subsequent queries including SELECTS will go to the primary for the duration of the request, as well as for subsequent requests from the same user up until `primary_ttl` seconds (I plan to make that 5)

File: app/controllers/application_controller.rb (hunk 2 above)
Comment (pleary) [reply app/controllers/application_controller.rb:757]: correct. There are some cases where we may need to write to the primary, but we know we won't ever read what was just written for the duration of the request, so there's no need to stick all subsequent queries to the primary

File: config/initializers/makara.rb (hunk 3 above)
Comment (pleary) [reply config/initializers/makara.rb:49]: `sql_skip_stickiness_matchers` is defined in the Makara library, and I'm overriding it here. These are patterns where, if the current query matches any of the patterns, it'll automatically ignore stickiness. I wanted to extend the library defaults w/o repeating the default values, so I'm extending them with the CUSTOM_* constants, and overriding the methods to use the extended matchers. I can add some more comments in this file

File: lib/elastic_model/acts_as_elastic_model.rb (hunk 4 above)
Comment (pleary) [reply lib/elastic_model/acts_as_elastic_model.rb:264]: When we index an obs, we'll update it's last_indexed_at time. But we pretty much never need to turn around and query for that attribute in a time frame longer than expected replication lag. So this is making sure that for any request that indexes an observation, it's not stuck querying the primary for the duration of the request after doing the indexing, so we can spread around the post-indexing query load. Where we stick and don't will need some tweaking as we see how this works and how well load is split between the DBs

**Comment by pleary** [comment]
Yes Chef will write database.yml, but for reference it'll look something like this:
```
production:
  username: USERNAME
  password: PASSWORD
  encoding: utf8
  database: DB_NAME
  sslmode: require
  adapter: makara_postgis
  makara:
    primary_ttl: 5
    sticky: true
    connections:
      - role: primary
        name: primary
        host: PRIMARY_HOST
        disable_blacklist: true
      - role: replica
        name: replica
        host: REPLICA_HOST_1
      ...
```

**Pull request merged**